test/e2e/expirable-airdrop.spec.ts

```diff
@@ -0,0 +1,250 @@
+import { BigNumber } from 'ethers';
+import { takeSnapshot, time, setBalance, impersonateAccount } from '@nomicfoundation/hardhat-network-helpers';
+import { MerkleTree } from 'merkletreejs';
+
+function getLeaf(address: string, amount: BigNumber) {
+  return Buffer.from(ethers.utils.solidityKeccak256(['address', 'uint256'], [address, amount]).slice(2), 'hex');
+}
+
+function createMerkleRoot(accounts: any[], amount: BigNumber) {
+  let toClaim = amount;
+
+  const leaves = accounts.map((x) => {
+    let address = x.address;
+    return getLeaf(address, toClaim);
+  });
+
+  const tree = new MerkleTree(leaves, ethers.utils.keccak256, { sortPairs: true });
+  const root = tree.getHexRoot();
+
+  return [root, tree];
+}
+
+describe('Expirable airdrop', () => {
+  let nowTimestamp: number;
+  const oneMonth: number = 30 * 24 * 60 * 60; // one month in seconds
+  let expirationTimestamp: number;
+  let accounts: any[];
+  let deployer: any;
+  let root: any;
+  let tree: any;
+  let token: any;
+  let expirableAirdrop: any;
+  let airdropAmount: number = 10;
+  let airdropAmountBN: BigNumber = toUnit(airdropAmount);
+  let snapshot: any;
+  let alice: SignerWithAddress;
+  let bob: SignerWithAddress;
+  let carl: SignerWithAddress;
+  let random: SignerWithAddress;
+  const randomAddress: string = '0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045';
+
+  async function prepareAirdrop() {
+    let toDeposit = toUnit(airdropAmount * (accounts.length + 1));
+    await token.mint(expirableAirdrop.address, toDeposit);
+  }
+
+  async function claim(sender: SignerWithAddress, claimee: SignerWithAddress) {
+    let leaf = getLeaf(claimee.address, airdropAmountBN);
+    let proof = tree.getHexProof(leaf);
+
+    await expirableAirdrop.connect(sender).claimAndSendToClaimee(claimee.address, airdropAmountBN, proof);
+  }
+
+  before(async () => {
+    nowTimestamp = await time.latest();
+    expirationTimestamp = nowTimestamp + 3 * oneMonth; // 3 months
+    [deployer, ...accounts] = await ethers.getSigners();
+
+    [root, tree] = createMerkleRoot(accounts, airdropAmountBN);
+
+    // deploy erc20 mock
+    const TokenFactory = await ethers.getContractFactory('ERC20Mock');
+    token = await TokenFactory.deploy('Token', 'TKN');
+
+    // deploy contract
+    const ExpirableAirdropFactory = await ethers.getContractFactory('ExpirableAirdrop');
+    expirableAirdrop = await ExpirableAirdropFactory.deploy(deployer.address, token.address, expirationTimestamp, root);
+
+    // load accounts
+    alice = accounts[1];
+    bob = accounts[2];
+    carl = accounts[3];
+    await impersonateAccount(randomAddress);
+    await setBalance(randomAddress, toUnit(10000));
+    random = await ethers.getSigner(randomAddress);
+
+    // save snapshot for later
+    snapshot = await takeSnapshot();
+  });
+
+  beforeEach(async () => {
+    await snapshot.restore();
+  });
+
+  when('deposit', () => {
```

Comment:
Could we test that the `Deposited` event is emmited?

I would suggest something like this:
```
when(’deposit is executed, () => {
    const toDeposit = airdropAmountBN.mul(accounts.length + 1)
    let tx: TransactionResponse
    given(() => {
        await token.mint(deployer.address, toDeposit);
        await token.approve(expirableAirdrop.address, ethers.constants.MaxUint256);
        tx = await expirableAirdrop.depositTokens(toDeposit);
    })
    then('tokens are deposited', () => {
        // Check that balance changed
    })
    then(‘event is emitted’, () => {
        expect(tx).to.have.emitted(…)
    }) 
})
```